src/main/kotlin/org/jetbrains/research/libsl/nodes/references/builders/TypeReferenceBuilder.kt

```diff
@@ -13,10 +16,59 @@ object TypeReferenceBuilder {
         isPointer: Boolean = false,
         context: LslContextBase
     ): TypeReference {
-        return TypeReference(name, isPointer, typeBound, genericReferences, context)
+        if (isLiteral(name))
+            return buildLiteralRef(name, context)
+        if (genericReferences.isNotEmpty())
+            return buildGenericRef(name, typeBound, genericReferences, context)
+        return buildPlainRef(name, isPointer, typeBound, context)
     }
 
-    fun Type.getReference(context: LslContextBase, typeBound: GenericTypeBound = GenericTypeBound.EMPTY): TypeReference {
-        return build(this.name, typeBound, this.generics, this.isPointer, context)
+    private fun buildLiteralRef(
+        name: String,
+        context: LslContextBase
+    ): TypeReference {
+        return LiteralTypeReference(name, context)
+    }
+
+    private fun buildGenericRef(
+        name: String,
+        typeBound: GenericTypeBound = GenericTypeBound.EMPTY,
+        genericReferences: MutableList<TypeReference>,
+        context: LslContextBase
+    ): TypeReference {
+        return GenericTypeReference(name, typeBound, genericReferences, context)
+    }
+
+    private fun buildPlainRef(
+        name: String,
+        isPointer: Boolean = false,
+        typeBound: GenericTypeBound = GenericTypeBound.EMPTY,
+        context: LslContextBase
+    ): TypeReference {
+        return PlainTypeReference(name, isPointer, typeBound, context)
+    }
+
+    fun Type.getReference(
+        context: LslContextBase,
+        typeBound: GenericTypeBound = GenericTypeBound.EMPTY
+    ): TypeReference {
+        if (isLiteral(this.name))
+            return buildLiteralRef(this.name, context)
+        if (this.generics.isNotEmpty())
+            return buildGenericRef(this.name, typeBound, this.generics, context)
+        return buildPlainRef(this.name, this.isPointer, typeBound, context)
+    }
+
+    fun isLiteral(name: String): Boolean {
```

Comment:
Shouldn't there be also an option for `null` value literal?
Example: `var foo: "something" | null = ...`

*Side-note: should we also introduce `unknown` somehow?*

Reply:
Fixed for null;

Yes, I agree with you about `unknown` type. I suggest to add this in the next pull request)